api/include/opentelemetry/trace/propagation/http_trace_context.h

```diff
@@ -0,0 +1,313 @@
+    uint8_t buf[kSpanIdBytes / 2];
+    int tmp;
+    for (int i = 0; i < kSpanIdBytes; i++)
+    {
+      tmp = CharToInt(*spn_id);
+      if (tmp < 0)
+      {
+        for (int j = 0; j < kSpanIdBytes / 2; j++)
+        {
+          buf[j] = 0;
+        }
+        return SpanId(buf);
+      }
+      if (i % 2 == 0)
+      {
+        buf[i / 2] = tmp * 16;
+      }
+      else
+      {
+        buf[i / 2] += tmp;
+      }
+      spn_id++;
+    }
+    return SpanId(buf);
+  }
+
+  static TraceFlags GenerateTraceFlagsFromString(nostd::string_view trace_flags)
+  {
+    uint8_t buf;
+    int tmp1 = CharToInt(trace_flags[0]);
+    int tmp2 = CharToInt(trace_flags[1]);
+    if (tmp1 < 0 || tmp2 < 0)
+      return TraceFlags(0);  // check for invalid char
+    buf = tmp1 * 16 + tmp2;
+    return TraceFlags(buf);
+  }
+
+private:
+  static uint8_t CharToInt(char c)
+  {
+    if (c >= '0' && c <= '9')
+    {
+      return (int)(c - '0');
+    }
+    else if (c >= 'a' && c <= 'f')
+    {
+      return (int)(c - 'a' + 10);
+    }
+    else if (c >= 'A' && c <= 'F')
+    {
+      return (int)(c - 'A' + 10);
+    }
+    else
+    {
+      return -1;
+    }
+  }
+
+  static void InjectTraceParent(const SpanContext &span_context, T &carrier, Setter setter)
+  {
+    char trace_id[32];
+    TraceId(span_context.trace_id()).ToLowerBase16(trace_id);
+    char span_id[16];
+    SpanId(span_context.span_id()).ToLowerBase16(span_id);
+    char trace_flags[2];
+    TraceFlags(span_context.trace_flags()).ToLowerBase16(trace_flags);
+    // Note: This is only temporary replacement for appendable string
+    std::string hex_string = "00-";
+    for (int i = 0; i < 32; i++)
+    {
+      hex_string += trace_id[i];
+    }
+    hex_string += "-";
+    for (int i = 0; i < 16; i++)
+    {
+      hex_string += span_id[i];
+    }
+    hex_string += "-";
+    for (int i = 0; i < 2; i++)
+    {
+      hex_string += trace_flags[i];
+    }
+    setter(carrier, kTraceParent, hex_string);
+  }
+
+  static void InjectImpl(Setter setter, T &carrier, const SpanContext &span_context)
+  {
+    InjectTraceParent(span_context, carrier, setter);
+  }
+
+  static SpanContext ExtractContextFromTraceParent(nostd::string_view trace_parent)
+  {
+    bool is_valid = trace_parent.length() == kHeaderSize && trace_parent[kVersionBytes] == '-' &&
+                    trace_parent[kVersionBytes + kTraceIdBytes + 1] == '-' &&
+                    trace_parent[kVersionBytes + kTraceIdBytes + kSpanIdBytes + 2] == '-';
+    if (!is_valid)
+    {
+      std::cout << "Unparseable trace_parent header. Returning INVALID span context." << std::endl;
+      return SpanContext(false, false);
+    }
+    nostd::string_view version;
+    nostd::string_view trace_id;
+    nostd::string_view span_id;
+    nostd::string_view trace_flags;
+    int elt_num   = 0;
+    int countdown = kHeaderElementLengths[elt_num];
+    int start_pos = -1;
+    for (int i = 0; i < int(trace_parent.size()); i++)
```

Comment:
Rather than iterating over `trace_parent` and tracking `start_pos` etc, how about using `trace_parent.substr()` based on the expected sizes? We know how long each segment (between '-' characters) is supposed to be, and we've used that to validated that the hyphens are in the right places.

After getting the parts, we can call a helper function one each one to make sure it contains the right set of characters (instead of the final else condition).

Reply:
I could, but the when I was trying to do the w3c tests I found that the rules sometime varies. And I am not sure if they are going to allow the existence of space or tabs in such strings and hence this way is safer and less vulnerable to change of valid id critera I believe.

Reply:
I've changed this to fixed size substr followed by a validation afterwards.